trieste/ask_tell_optimization.py

```diff
@@ -421,18 +437,44 @@ def tell(self, new_data: Mapping[Tag, Dataset] | Dataset) -> None:
         if isinstance(new_data, Dataset):
             new_data = {OBJECTIVE: new_data}
 
-        if self._datasets.keys() != new_data.keys():
+        # The datasets must have the same keys as the existing datasets. Only exception is if
+        # the existing datasets are all global, in which case the dataset will be appropriately
+        # updated below for the next iteration.
+        datasets_indices = {LocalizedTag.from_tag(tag).local_index for tag in self._datasets.keys()}
+        if self._datasets.keys() != new_data.keys() and datasets_indices != {None}:
             raise ValueError(
                 f"new_data keys {new_data.keys()} doesn't "
                 f"match dataset keys {self._datasets.keys()}"
             )
 
-        for tag in self._datasets:
-            self._datasets[tag] += new_data[tag]
+        # In order to support local datasets, account for the case where there may be an initial
+        # dataset that is not tagged per region. In this case, only the global dataset will exist
+        # in datasets. We want to copy this initial dataset to all the regions.
+        # If a tag from tagged_output does not exist in datasets, then add it to
+        # datasets by copying the data from datasets with the same global tag. Otherwise keep the
+        # existing data from datasets.
+        #
+        # Note: this replication of initial data can potentially cause an issue when a global model
+        # is being used with local datasets, as the points may be repeated. This will only be an
+        # issue if two regions overlap and both contain that initial data-point -- as filtering
+        # (in BatchTrustRegion) would otherwise remove duplicates. The main way to avoid the issue
+        # in this scenario is to provide local initial datasets, instead of a global initial
+        # dataset.
+        updated_datasets = {}
+        for tag, new_dataset in new_data.items():
+            _, old_dataset = get_value_for_tag(
```

Comment:
Would it be better (clearer but also more efficient) to reuse rather than copy the dataset if it is already correctly tagged?

```python
for tag, new_dataset in new_data.items():
    if tag in self._datasets:
        self._datasets[tag] += new_dataset
    else:
        global_tag = LocalizedTag.from_tag(tag).global_tag
        assert global_tag in self._datasets
        self._datasets[tag] = self._datasets[global_tag] + new_data
```